### PR TITLE
update `quat(::Array)` to return `Array{<:Quaternion}`

### LINKED ISSUE #94: `quat([1,2,3,4])` should return `Vector{Quaternion{Int}}`
The function `quat` should be similar to `complex`, but the behavior is different.

```julia
julia> complex(1)
1 + 0im

julia> quat(1)
Quaternion{Int64}(1, 0, 0, 0, true)

julia> complex([1,2,3,4])
4-element Vector{Complex{Int64}}:
 1 + 0im
 2 + 0im
 3 + 0im
 4 + 0im

julia> quat([1,2,3,4])
Quaternion{Int64}(0, 1, 2, 3, false)
```

I think `quat([1,2,3,4])` should be `[Quaternion{Int64}(1,0,0,0), Quaternion{Int64}(2,0,0,0), Quaternion{Int64}(3,0,0,0), Quaternion{Int64}(4,0,0,0)]`.

### PR DESCRIPTION
**Pull request by hyrodium**
This PR fixes #94.
Note that this PR is considered as a breaking change.

https://github.com/JuliaGeometry/Quaternions.jl/blob/0be4cdd881872e14b480bdb69326f188f4b40236/test/Quaternion.jl#L112-L113

### CODE REVIEW

File: src/Quaternion.jl [hunk 1]
```diff
@@ -24,10 +24,33 @@ promote_rule(::Type{Quaternion{T}}, ::Type{S}) where {T <: Real, S <: Real} = Qu
...
 quat(p, v1, v2, v3) = Quaternion(p, v1, v2, v3)
 quat(p, v1, v2, v3, n) = Quaternion(p, v1, v2, v3, n)
 quat(x) = Quaternion(x)
 quat(s, a) = Quaternion(s, a)
+function quat(A::AbstractArray{T}) where T
+    if !isconcretetype(T)
+        error("`quat` not defined on abstractly-typed arrays; please convert to a more specific type")
+    end
+    convert(AbstractArray{typeof(quat(zero(T)))}, A)
+end
```
Comment (hyrodium) [inline src/Quaternion.jl:53]: This implementation is the same as 
https://github.com/JuliaLang/julia/blob/1ece3f64e093ba47bcd5e7a01193b8c9b05d441e/base/complex.jl#L1114-L1119

**Comment by hyrodium** [comment]
I'm keeping this PR draft because I'm not sure whether the `quat` behavior in #94 is intended or not.

**Comment by hyrodium** [comment]
I'll remove some methods such as 
```julia
function Quaternion(s::Real, a::AbstractVector)
    Base.depwarn("`Quaternion(s::Real, a::AbstractVector)` is deprecated and will be removed in the next breaking release (v0.7.0). Please use `Quaternion(s, a[1], a[2], a[3])` instead.", :Quaternion)
    Quaternion(s, a[1], a[2], a[3])
end
```
after merging #111 and #113 to avoid making conflicts.

File: src/Quaternion.jl [hunk 2]
```diff
@@ -48,15 +40,24 @@ Quaternion{Int64}(7, 0, 0, 0)
...
+ Quaternion{Int64}(1, 0, 0, 0)
+ Quaternion{Int64}(2, 0, 0, 0)
+ Quaternion{Int64}(3, 0, 0, 0)
 ```
 """
 quat
 
 quat(p, v1, v2, v3) = Quaternion(p, v1, v2, v3)
 quat(x) = Quaternion(x)
 quat(s, a) = Quaternion(s, a)
```
Comment (sethaxen) [inline src/Quaternion.jl:54]: This method now leads nowhere and should also be deleted.
```suggestion
```

File: src/Quaternion.jl [hunk 3]
```diff
@@ -48,15 +40,24 @@ Quaternion{Int64}(7, 0, 0, 0)
...
+function quat(A::AbstractArray{T}) where T
+    if !isconcretetype(T)
+        error("`quat` not defined on abstractly-typed arrays; please convert to a more specific type")
```
Comment (sethaxen) [inline src/Quaternion.jl:57]: Can you add a test for this error?

**Review by sethaxen — CHANGES_REQUESTED** [review]
LGTM, just 2 fixes needed.

File: src/Quaternion.jl (hunk 3 above)
Comment (hyrodium) [reply src/Quaternion.jl:56]: Done!

**Review by sethaxen — APPROVED** [review]
LGTM!

**Pull request merged**